File: src/scpn_fusion/core/rf_heating.py
# ──────────────────────────────────────────────────────────────────────
# SCPN Fusion Core — RF Heating
# © 1998–2026 Miroslav Šotek. All rights reserved.
# Contact: www.anulum.li | protoscience@anulum.li
# ORCID: https://orcid.org/0009-0009-3560-0851
# License: GNU AGPL v3 | Commercial licensing available
# ──────────────────────────────────────────────────────────────────────
from __future__ import annotations

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import odeint
import sys
import os

try:
    from scpn_fusion.core._rust_compat import FusionKernel
except ImportError:
    from scpn_fusion.core.fusion_kernel import FusionKernel


def _require_finite_float(
    name: str,
    value: float,
    *,
    min_value: float | None = None,
    max_value: float | None = None,
) -> float:
    out = float(value)
    if not np.isfinite(out):
        raise ValueError(f"{name} must be finite.")
    if min_value is not None and out < min_value:
        raise ValueError(f"{name} must be >= {min_value}.")
    if max_value is not None and out > max_value:
        raise ValueError(f"{name} must be <= {max_value}.")
    return out


def _require_int(name: str, value: int, minimum: int) -> int:
    if isinstance(value, bool) or not isinstance(value, (int, np.integer)):
        raise ValueError(f"{name} must be an integer >= {minimum}.")
    out = int(value)
    if out < minimum:
        raise ValueError(f"{name} must be an integer >= {minimum}.")
    return out

class RFHeatingSystem:
    """
    Simulates Ion Cyclotron Resonance Heating (ICRH).
    Uses Ray-Tracing to track EM waves launching from the antenna 
    and absorbing at the resonance layer.
    """
    def __init__(self, config_path):
        self.kernel = FusionKernel(config_path)
        self.kernel.solve_equilibrium() # Get B-field map
        
        # Physics Constants
        self.q_D = 1.602e-19 # Charge (Deuterium)
        self.m_D = 3.34e-27  # Mass (Deuterium)
        self.freq = 50e6     # 50 MHz (Standard ICRH freq)
        self.omega_wave = 2 * np.pi * self.freq
        
    def get_plasma_params(self, R, Z):
        """
        Returns B_mod, density, and derivatives at (R,Z).
        """
        # 1. Magnetic Field (Toroidal dominates)
        # B_tor ~ B0 * R0 / R
        B0 = 5.3 # Tesla at axis (ITER)
        R0 = 6.2
        B_tor = B0 * R0 / R
        
        # Poloidal field from kernel
        # We need grid lookup
        ir = int((R - self.kernel.R[0]) / self.kernel.dR)
        iz = int((Z - self.kernel.Z[0]) / self.kernel.dZ)
        
        if 0 <= ir < self.kernel.NR and 0 <= iz < self.kernel.NZ:
            B_R = self.kernel.B_R[iz, ir]
            B_Z = self.kernel.B_Z[iz, ir]
            psi_val = self.kernel.Psi[iz, ir]
        else:
            B_R, B_Z, psi_val = 0, 0, 0
            
        B_mod = np.sqrt(B_tor**2 + B_R**2 + B_Z**2)
        
        # 2. Density Profile (Parabolic)
        # n = n0 * (1 - psi_norm)
        # Reduced-order surrogate: Gaussian blob
        dist_sq = (R - R0)**2 + Z**2
        n_e = 1e20 * np.exp(-dist_sq / 2.0)
        
        # Derivatives of density (for refraction)
        dn_dR = -n_e * (R - R0) / 1.0
        dn_dZ = -n_e * Z / 1.0
        
        return B_mod, n_e, dn_dR, dn_dZ

    def dispersion_relation(self, R, Z, k_R, k_Z):
        """
        Calculates the local dispersion D(w, k) = 0.
        Harden with Warm Plasma thermal corrections for ICRH.
        """
        B_mod, n_e, _, _ = self.get_plasma_params(R, Z)
        
        if n_e < 1e18: return 1.0 # Vacuum
        
        # 1. Alfven speed (Cold Limit)
        mu0 = 4*np.pi*1e-7
        v_A = B_mod / np.sqrt(mu0 * n_e * self.m_D)
        
        # 2. Thermal correction (Warm Plasma)
        # For ICRH, thermal effects modify the k_perp dispersion
        T_ion_keV = 10.0 # Heuristic local Ti
        v_thi = np.sqrt(2.0 * T_ion_keV * 1.602e-16 / self.m_D)
        
        # Finite Larmor Radius (FLR) correction factor
        # omega^2 = k^2 * v_A^2 * (1 + 3/4 * (k_perp * rho_i)^2)
        rho_i = self.m_D * v_thi / (self.q_D * B_mod)
        k_sq = k_R**2 + k_Z**2
        
        flr_correction = 1.0 + 0.75 * (k_sq * rho_i**2)
        
        # Dispersion: D = k^2 * v_A^2 * flr_correction - omega^2
        D = k_sq * v_A**2 * flr_correction - self.omega_wave**2
        
        return D

    def ray_equations(self, state, t):
        """
        Hamiltonian Ray Tracing equations.
        dr/dt = dD/dk
        dk/dt = -dD/dr
        """
        R, Z, kR, kZ = state
        
        # Finite differences for derivatives of D
        # This implicitly handles refraction and reflection
        eps = 1e-3
        
        # dD/dk
        D_pkR = self.dispersion_relation(R, Z, kR + eps, kZ)
        D_mkR = self.dispersion_relation(R, Z, kR - eps, kZ)
        dD_dkR = (D_pkR - D_mkR) / (2*eps)
        
        D_pkZ = self.dispersion_relation(R, Z, kR, kZ + eps)
        D_mkZ = self.dispersion_relation(R, Z, kR, kZ - eps)
        dD_dkZ = (D_pkZ - D_mkZ) / (2*eps)
        
        # dD/dr
        D_pR = self.dispersion_relation(R + eps, Z, kR, kZ)
        D_mR = self.dispersion_relation(R - eps, Z, kR, kZ)
        dD_dR = (D_pR - D_mR) / (2*eps)
        
        D_pZ = self.dispersion_relation(R, Z + eps, kR, kZ)
        D_mZ = self.dispersion_relation(R, Z - eps, kR, kZ)
        dD_dZ = (D_pZ - D_mZ) / (2*eps)
        
        # Group Velocity (dr/dt)
        dR_dt = -dD_dkR
        dZ_dt = -dD_dkZ
        
        # Wavevector change (dk/dt)
        dkR_dt = dD_dR
        dkZ_dt = dD_dZ
        
        return [dR_dt, dZ_dt, dkR_dt, dkZ_dt]

    def trace_rays(self, n_rays=10):
        print("--- RF HEATING RAY TRACING ---")
        print(f"Frequency: {self.freq/1e6} MHz")
        
        # Antenna Position (Outboard midplane)
        R_ant = 9.0
        Z_ant_spread = np.linspace(-1.0, 1.0, n_rays)
        
        trajectories = []
        
        for i in range(n_rays):
            # Initial condition: Launch inward (kR < 0)
            k0 = 10.0 # Initial wavenumber
            init_state = [R_ant, Z_ant_spread[i], -k0, 0.0]
            
            t_span = np.linspace(0, 0.5, 100) # Short time (normalized)
            
            # Solve ODE
            sol = odeint(self.ray_equations, init_state, t_span)
            
            # Check Resonance
            # Resonance condition: omega = omega_ci = qB/m
            # B_res = omega * m / q
            B_res = self.omega_wave * self.m_D / self.q_D
            
            # Find where B matches B_res along path
            # (Post-processing check)
            
            trajectories.append(sol)
            
        print(f"Resonance Field B_res: {B_res:.2f} Tesla")
        return trajectories, B_res

    def plot_heating(self, trajectories, B_res):
        fig, ax = plt.subplots(figsize=(8, 8))
        
        # 1. Plasma Geometry
        ax.contour(self.kernel.RR, self.kernel.ZZ, self.kernel.Psi, colors='gray', alpha=0.3)
        
        # 2. Resonance Layer (Vertical Line approx)
        # B_tor ~ B0*R0/R. B ~ B_res => R_res ~ B0*R0/B_res
        R_res = (5.3 * 6.2) / B_res
        ax.axvline(R_res, color='green', linestyle='--', linewidth=2, label='Cyclotron Resonance Layer')
        
        # 3. Rays
        for i, sol in enumerate(trajectories):
            R = sol[:, 0]
            Z = sol[:, 1]
            ax.plot(R, Z, 'r-', alpha=0.6)
            
            # Draw absorption point (intersection with Resonance)
            # Simple geometric check
            idx = np.abs(R - R_res).argmin()
            if np.abs(R[idx] - R_res) < 0.1:
                ax.plot(R[idx], Z[idx], 'y*', markersize=10, label='Energy Dump' if i==0 else "")
        
        # Antenna
        ax.plot([9.0]*len(trajectories), [t[0,1] for t in trajectories], 'ko', label='Antenna Array')
        
        ax.set_title(f"ICRH Wave Propagation ({self.freq/1e6} MHz)")
        ax.set_xlabel("R (m)")
        ax.set_ylabel("Z (m)")
        ax.set_xlim(2, 10)
        ax.set_ylim(-6, 6)
        ax.legend()
        
        plt.savefig("RF_Heating_Rays.png")
        print("Saved: RF_Heating_Rays.png")

    def compute_power_deposition(self, trajectories, P_rf_mw=20.0, n_radial_bins=50):
        """Compute radial power deposition profile from ray absorption.

        Uses cyclotron damping: the imaginary part of the wave vector causes
        exponential power decay along each ray path. The absorption coefficient
        is proportional to exp(-(omega - n*Omega_ci)^2 / (k_par * v_thi)^2).

        Parameters
        ----------
        trajectories : list of ndarray
            Ray paths from trace_rays().
        P_rf_mw : float
            Total injected RF power (MW).
        n_radial_bins : int
            Number of radial bins for the deposition profile.

        Returns
        -------
        rho_bins : ndarray
            Normalised radius bin centres (0–1).
        P_dep_mw_m3 : ndarray
            Power deposition density (MW/m^3) per radial bin.
        absorption_efficiency : float
            Fraction of injected power absorbed (0–1).
        """
        R0 = 6.2
        a = 2.0  # minor radius (m)
        B0 = 5.3
        T_ion_keV = 20.0

        # Ion thermal speed
        v_thi = np.sqrt(2.0 * T_ion_keV * 1e3 * 1.602e-19 / self.m_D)

        rho_bins = np.linspace(0.0, 1.0, n_radial_bins)
        P_dep = np.zeros(n_radial_bins)
        total_absorbed = 0.0
        P_per_ray = P_rf_mw / max(len(trajectories), 1)

        for sol in trajectories:
            R = sol[:, 0]
            Z = sol[:, 1]
            P_ray = P_per_ray  # power remaining in this ray (MW)

            for j in range(1, len(R)):
                if P_ray < 1e-6:
                    break

                R_mid = 0.5 * (R[j - 1] + R[j])
                Z_mid = 0.5 * (Z[j - 1] + Z[j])
                ds = np.sqrt((R[j] - R[j - 1]) ** 2 + (Z[j] - Z[j - 1]) ** 2)

                if ds < 1e-12:
                    continue

                # Local B and cyclotron frequency
                B_local = B0 * R0 / max(R_mid, 0.1)
                omega_ci = self.q_D * B_local / self.m_D

                # Cyclotron damping coefficient
                # k_imag ~ (omega_pi^2 / (c * omega)) * exp(-delta^2)
                # where delta = (omega - omega_ci) / (k_par * v_thi)
                delta = (self.omega_wave - omega_ci) / max(10.0 * v_thi, 1e6)
                damping = np.exp(-delta**2)

                # Absorption: dP = -alpha * P * ds
                # alpha ~ 0.5 * damping / a  (normalised so that strong resonance
                # absorbs over ~1 minor radius length scale)
                alpha = 0.5 * damping / max(a, 0.1)
                dP = P_ray * (1.0 - np.exp(-alpha * ds))
                P_ray -= dP
                total_absorbed += dP

                # Map to radial bin
                rho = np.sqrt((R_mid - R0) ** 2 + Z_mid**2) / a
                rho = min(rho, 1.0)
                bin_idx = min(int(rho * n_radial_bins), n_radial_bins - 1)

                # Volume of radial shell
                dr = 1.0 / n_radial_bins
                r_inner = rho_bins[bin_idx] * a
                dV = 2.0 * np.pi * R_mid * 2.0 * np.pi * r_inner * a * dr
                dV = max(dV, 1e-6)

                P_dep[bin_idx] += dP / dV  # MW / m^3

        efficiency = total_absorbed / max(P_rf_mw, 1e-12)
        return rho_bins, P_dep, float(np.clip(efficiency, 0.0, 1.0))


class ECRHHeatingSystem:
    """Electron Cyclotron Resonance Heating (ECRH) with ray-tracing absorption.

    Operates at 170 GHz (ITER ECRH frequency). Resonance occurs where
    omega = n * Omega_ce (electron cyclotron frequency), with n=1 (fundamental)
    or n=2 (second harmonic).

    Parameters
    ----------
    b0_tesla : float
        On-axis toroidal magnetic field (T).
    r0_major : float
        Major radius (m).
    freq_ghz : float
        ECRH frequency (GHz). Default: 170.
    harmonic : int
        Cyclotron harmonic number (1 or 2).
    """

    def __init__(
        self,
        b0_tesla: float = 5.3,
        r0_major: float = 6.2,
        freq_ghz: float = 170.0,
        harmonic: int = 1,
    ):
        self.B0 = _require_finite_float("b0_tesla", b0_tesla, min_value=0.1)
        self.R0 = _require_finite_float("r0_major", r0_major, min_value=0.1)
        self.freq = _require_finite_float("freq_ghz", freq_ghz, min_value=0.1) * 1e9
        self.omega = 2.0 * np.pi * self.freq
        self.harmonic = _require_int("harmonic", harmonic, 1)
        self.m_e = 9.109e-31
        self.q_e = 1.602e-19

    def resonance_radius(self) -> float:
        """Major radius where n*Omega_ce = omega."""
        B_res = self.omega * self.m_e / (self.harmonic * self.q_e)
        return self.B0 * self.R0 / B_res

    def compute_deposition(
        self,
        P_ecrh_mw: float = 20.0,
        n_radial_bins: int = 50,
        T_e_keV: float = 20.0,
        n_e: float = 1e20,
        launch_angle_deg: float = 0.0,
    ) -> tuple[np.ndarray, np.ndarray, float]:
        """Compute ECRH power deposition profile.

        Uses Gaussian deposition centred at the resonance layer with width
        determined by the Doppler broadening of the electron cyclotron resonance.

        Returns
        -------
        rho_bins : ndarray
            Normalised radius (0–1).
        P_dep_mw_m3 : ndarray
            Power deposition density.
        absorption_efficiency : float
            Fraction absorbed (0–1).
        """
        P_ecrh_mw = _require_finite_float("P_ecrh_mw", P_ecrh_mw, min_value=0.0)
        n_radial_bins = _require_int("n_radial_bins", n_radial_bins, 8)
        T_e_keV = _require_finite_float("T_e_keV", T_e_keV, min_value=0.01)
        n_e = _require_finite_float("n_e", n_e, min_value=1e16)
        launch_angle_deg = _require_finite_float(
            "launch_angle_deg",
            launch_angle_deg,
            min_value=-85.0,
            max_value=85.0,
        )

        a = 2.0  # minor radius
        R_res = self.resonance_radius()
        rho_res = abs(R_res - self.R0) / a

        # Thermal speed for Doppler width
        v_the = np.sqrt(2.0 * T_e_keV * 1e3 * self.q_e / self.m_e)
        theta = np.deg2rad(launch_angle_deg)
        obliquity = float(np.clip(np.cos(theta) ** 2, 0.05, 1.0))
        # Doppler width in rho: delta_rho ~ v_the / (omega * a)
        delta_rho = max(v_the / (self.omega * a) * 50.0 * (1.0 + 0.35 * abs(np.sin(theta))), 0.02)

        rho_bins = np.linspace(0.0, 1.0, n_radial_bins)
        P_dep = np.zeros(n_radial_bins)

        # Gaussian deposition profile centred at rho_res
        for i, rho in enumerate(rho_bins):
            r_shell = rho * a
            R_local = self.R0 + r_shell
            dV = 2.0 * np.pi * R_local * 2.0 * np.pi * r_shell * a / n_radial_bins
            dV = max(dV, 1e-6)
            dep = np.exp(-((rho - rho_res) ** 2) / (2.0 * delta_rho**2))
            P_dep[i] = dep / dV

        # Normalise so total = P_ecrh * absorption
        # Single-pass absorption: eta ~ 1 - exp(-tau_opt)
        # Optical depth for O-mode: tau ~ (omega_pe / omega)^2 * (n_e * L / ...)
        omega_pe = np.sqrt(n_e * self.q_e**2 / (self.m_e * 8.854e-12))
        resonance_overlap = 1.0 if rho_res <= 1.0 else float(np.exp(-((rho_res - 1.0) / 0.18) ** 2))
        tau_opt = (omega_pe / self.omega) ** 2 * 20.0 * self.harmonic * obliquity * resonance_overlap
        efficiency = float(np.clip(1.0 - np.exp(-tau_opt), 0.01, 0.99))

        total_dep = np.sum(P_dep)
        if total_dep > 1e-12:
            P_dep *= (P_ecrh_mw * efficiency) / total_dep

        return rho_bins, P_dep, efficiency


if __name__ == "__main__":
    cfg = "03_CODE/SCPN-Fusion-Core/validation/iter_validated_config.json"
    rf = RFHeatingSystem(cfg)
    rays, B_res = rf.trace_rays()
    rf.plot_heating(rays, B_res)

File: scpn-fusion-rs/crates/fusion-core/src/rf_heating.rs
// ─────────────────────────────────────────────────────────────────────
// SCPN Fusion Core — RF Heating
// © 1998–2026 Miroslav Šotek. All rights reserved.
// Contact: www.anulum.li | protoscience@anulum.li
// ORCID: https://orcid.org/0009-0009-3560-0851
// License: GNU AGPL v3 | Commercial licensing available
// ─────────────────────────────────────────────────────────────────────
//! ICRH ray tracing for RF heating deposition.
//!
//! Port of `rf_heating.py` (200 lines).
//! Uses Hamiltonian ray equations in cold plasma approximation.

use fusion_types::constants::MU0_SI;
use fusion_types::error::{FusionError, FusionResult};

/// Deuterium charge [C]. Python line 23.
const Q_D: f64 = 1.602e-19;

/// Deuterium mass [kg]. Python line 24.
const M_D: f64 = 3.34e-27;

/// ICRH frequency [Hz]. Python line 25.
const FREQ_HZ: f64 = 50e6;

/// Angular frequency [rad/s].
const OMEGA_WAVE: f64 = 2.0 * std::f64::consts::PI * FREQ_HZ;

/// Magnetic field on axis [T]. Python line 34.
const B0: f64 = 5.3;

/// Major radius [m]. Python line 35.
const R0: f64 = 6.2;

/// Peak electron density [m⁻³]. Python line 56.
const N_E_PEAK: f64 = 1.0e20;

/// Density Gaussian width [m²]. Python line 56.
const DENSITY_WIDTH: f64 = 2.0;

/// Finite difference step for gradient [m]. Python line 96.
const GRAD_EPS: f64 = 1e-3;

/// Initial wavenumber [m⁻¹]. Python line 139.
const K0_INITIAL: f64 = 10.0;

/// Number of rays to trace. Python line 131.
const N_RAYS: usize = 10;

/// Antenna radial position [m]. Python line 131.
const R_ANTENNA: f64 = 9.0;

/// Resonance layer intersection tolerance [m]. Python line 179.
const RESONANCE_TOL: f64 = 0.1;

/// Ray state: (R, Z, k_R, k_Z).
pub type RayState = [f64; 4];

/// Result of a single ray trace.
#[derive(Debug, Clone)]
pub struct RayTraceResult {
    /// Ray trajectory points (R, Z).
    pub trajectory: Vec<(f64, f64)>,
    /// Resonance crossing position, if found.
    pub resonance_point: Option<(f64, f64)>,
    /// Resonance layer R coordinate.
    pub r_resonance: f64,
}

/// ICRH ray tracing system.
pub struct RFHeatingSystem {
    pub omega: f64,
    pub b0: f64,
    pub r0: f64,
}

impl RFHeatingSystem {
    /// Create with ITER-like parameters.
    pub fn new() -> Self {
        RFHeatingSystem {
            omega: OMEGA_WAVE,
            b0: B0,
            r0: R0,
        }
    }

    fn validate_runtime_parameters(&self) -> FusionResult<()> {
        if !self.omega.is_finite() || self.omega <= 0.0 {
            return Err(FusionError::ConfigError(format!(
                "RF omega must be finite and > 0, got {}",
                self.omega
            )));
        }
        if !self.b0.is_finite() || self.b0 <= 0.0 {
            return Err(FusionError::ConfigError(format!(
                "RF b0 must be finite and > 0, got {}",
                self.b0
            )));
        }
        if !self.r0.is_finite() || self.r0 <= 0.0 {
            return Err(FusionError::ConfigError(format!(
                "RF r0 must be finite and > 0, got {}",
                self.r0
            )));
        }
        Ok(())
    }

    /// Toroidal magnetic field at radius R: B(R) = B0 · R0 / R.
    fn b_field(&self, r: f64) -> f64 {
        self.b0 * self.r0 / r
    }

    /// Electron density profile (Gaussian). Python line 56.
    fn density(&self, r: f64, z: f64) -> f64 {
        let dist_sq = (r - self.r0).powi(2) + z.powi(2);
        N_E_PEAK * (-dist_sq / DENSITY_WIDTH).exp()
    }

    /// Alfvén speed: v_A = B / sqrt(μ₀ n_e m_D).
    fn alfven_speed(&self, r: f64, z: f64) -> f64 {
        let b = self.b_field(r);
        let n = self.density(r, z).max(1e10); // floor to avoid division by zero
        b / (MU0_SI * n * M_D).sqrt()
    }

    /// Dispersion relation: D = k²v_A² - ω².
    fn dispersion(&self, r: f64, z: f64, kr: f64, kz: f64) -> f64 {
        let k_sq = kr * kr + kz * kz;
        let v_a = self.alfven_speed(r, z);
        k_sq * v_a * v_a - self.omega * self.omega
    }

    /// Ray equations (Hamiltonian): ds/dt = (dR/dt, dZ/dt, dk_R/dt, dk_Z/dt).
    ///
    /// dR/dt = -∂D/∂k_R,  dZ/dt = -∂D/∂k_Z
    /// dk_R/dt = ∂D/∂R,   dk_Z/dt = ∂D/∂Z
    fn ray_rhs(&self, state: &RayState) -> RayState {
        let [r, z, kr, kz] = *state;
        let eps = GRAD_EPS;

        // ∂D/∂k_R, ∂D/∂k_Z via finite differences
        let dd_dkr = (self.dispersion(r, z, kr + eps, kz) - self.dispersion(r, z, kr - eps, kz))
            / (2.0 * eps);
        let dd_dkz = (self.dispersion(r, z, kr, kz + eps) - self.dispersion(r, z, kr, kz - eps))
            / (2.0 * eps);

        // ∂D/∂R, ∂D/∂Z via finite differences
        let dd_dr = (self.dispersion(r + eps, z, kr, kz) - self.dispersion(r - eps, z, kr, kz))
            / (2.0 * eps);
        let dd_dz = (self.dispersion(r, z + eps, kr, kz) - self.dispersion(r, z - eps, kr, kz))
            / (2.0 * eps);

        [
            -dd_dkr, // dR/dt
            -dd_dkz, // dZ/dt
            dd_dr,   // dk_R/dt
            dd_dz,   // dk_Z/dt
        ]
    }

    /// Adaptive RK4 integration step with sub-stepping.
    /// The Alfvén speed can be ~1e7 m/s, so group velocity dR/dt ~ k·v_A²
    /// can be extremely large. We use adaptive sub-stepping to keep
    /// displacement per step < 0.1 m.
    fn rk4_step(&self, state: &RayState, dt: f64) -> RayState {
        let k1 = self.ray_rhs(state);

        let s2 = [
            state[0] + 0.5 * dt * k1[0],
            state[1] + 0.5 * dt * k1[1],
            state[2] + 0.5 * dt * k1[2],
            state[3] + 0.5 * dt * k1[3],
        ];
        let k2 = self.ray_rhs(&s2);

        let s3 = [
            state[0] + 0.5 * dt * k2[0],
            state[1] + 0.5 * dt * k2[1],
            state[2] + 0.5 * dt * k2[2],
            state[3] + 0.5 * dt * k2[3],
        ];
        let k3 = self.ray_rhs(&s3);

        let s4 = [
            state[0] + dt * k3[0],
            state[1] + dt * k3[1],
            state[2] + dt * k3[2],
            state[3] + dt * k3[3],
        ];
        let k4 = self.ray_rhs(&s4);

        [
            state[0] + (dt / 6.0) * (k1[0] + 2.0 * k2[0] + 2.0 * k3[0] + k4[0]),
            state[1] + (dt / 6.0) * (k1[1] + 2.0 * k2[1] + 2.0 * k3[1] + k4[1]),
            state[2] + (dt / 6.0) * (k1[2] + 2.0 * k2[2] + 2.0 * k3[2] + k4[2]),
            state[3] + (dt / 6.0) * (k1[3] + 2.0 * k2[3] + 2.0 * k3[3] + k4[3]),
        ]
    }

    /// Ion cyclotron resonance radius.
    ///
    /// ω_ci = qB/m → B_res = ωm/q → R_res = B0·R0/B_res
    pub fn resonance_radius(&self) -> FusionResult<f64> {
        self.validate_runtime_parameters()?;
        let b_res = self.omega * M_D / Q_D;
        if !b_res.is_finite() || b_res <= 0.0 {
            return Err(FusionError::ConfigError(
                "RF resonance field must be finite and > 0".to_string(),
            ));
        }
        let r_res = self.b0 * self.r0 / b_res;
        if !r_res.is_finite() || r_res <= 0.0 {
            return Err(FusionError::ConfigError(
                "RF resonance radius became non-finite or non-positive".to_string(),
            ));
        }
        Ok(r_res)
    }

    /// Trace a single ray from given initial conditions.
    ///
    /// Uses adaptive time stepping: estimates RHS magnitude and scales dt
    /// so the spatial displacement per step is bounded.
    pub fn trace_single_ray(
        &self,
        r0: f64,
        z0: f64,
        kr0: f64,
        kz0: f64,
        n_steps: usize,
        _dt_hint: f64,
    ) -> FusionResult<RayTraceResult> {
        self.validate_runtime_parameters()?;
        if !r0.is_finite() || r0 <= 0.0 || !z0.is_finite() || !kr0.is_finite() || !kz0.is_finite() {
            return Err(FusionError::ConfigError(
                "RF ray initial conditions must be finite with r0 > 0".to_string(),
            ));
        }
        if n_steps == 0 {
            return Err(FusionError::ConfigError(
                "RF ray tracing requires n_steps >= 1".to_string(),
            ));
        }

        let r_res = self.resonance_radius()?;
        let mut state: RayState = [r0, z0, kr0, kz0];
        let mut trajectory = Vec::with_capacity(n_steps);
        let mut resonance_point = None;

        // Maximum spatial displacement per step [m]
        const MAX_DISP: f64 = 0.05;

        for _ in 0..n_steps {
            if state.iter().any(|v| !v.is_finite()) {
                return Err(FusionError::ConfigError(
                    "RF ray state became non-finite".to_string(),
                ));
            }
            trajectory.push((state[0], state[1]));

            // Check resonance crossing
            if resonance_point.is_none() && (state[0] - r_res).abs() < RESONANCE_TOL {
                resonance_point = Some((state[0], state[1]));
            }

            // Adaptive dt: estimate RHS magnitude
            let rhs = self.ray_rhs(&state);
            if rhs.iter().any(|v| !v.is_finite()) {
                return Err(FusionError::ConfigError(
                    "RF ray RHS became non-finite".to_string(),
                ));
            }
            let v_mag = (rhs[0] * rhs[0] + rhs[1] * rhs[1]).sqrt();
            if !v_mag.is_finite() {
                return Err(FusionError::ConfigError(
                    "RF ray velocity magnitude became non-finite".to_string(),
                ));
            }
            let dt = if v_mag > 1e-10 {
                MAX_DISP / v_mag
            } else {
                1e-3
            };
            if !dt.is_finite() || dt <= 0.0 {
                return Err(FusionError::ConfigError(
                    "RF adaptive timestep became invalid".to_string(),
                ));
            }

            state = self.rk4_step(&state, dt);

            // Boundary check (stay within reasonable domain)
            if state[0] < 0.5 || state[0] > 15.0 || state[1].abs() > 10.0 {
                break;
            }
        }

        Ok(RayTraceResult {
            trajectory,
            resonance_point,
            r_resonance: r_res,
        })
    }

    /// Trace all rays from the antenna array. Python lines 126-157.
    pub fn trace_rays(&self) -> FusionResult<Vec<RayTraceResult>> {
        self.validate_runtime_parameters()?;
        let n_steps = 500;
        let dt = 0.0; // adaptive

        let mut results = Vec::with_capacity(N_RAYS);
        for i in 0..N_RAYS {
            let z_antenna = -1.0 + 2.0 * (i as f64) / ((N_RAYS - 1) as f64);
            let result =
                self.trace_single_ray(R_ANTENNA, z_antenna, -K0_INITIAL, 0.0, n_steps, dt)?;
            results.push(result);
        }
        Ok(results)
    }
}

impl Default for RFHeatingSystem {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_resonance_radius() {
        let rf = RFHeatingSystem::new();
        let r_res = rf
            .resonance_radius()
            .expect("valid RF resonance-radius parameters");
        // For ITER: B0=5.3T, R0=6.2m, f=50MHz
        // B_res = 2π·50e6·3.34e-27/1.602e-19 ≈ 6.54 T
        // R_res = 5.3·6.2/6.54 ≈ 5.02 m
        assert!(r_res > 3.0 && r_res < 8.0, "Resonance radius: {r_res}");
    }

    #[test]
    fn test_trace_single_ray() {
        let rf = RFHeatingSystem::new();
        let result = rf
            .trace_single_ray(R_ANTENNA, 0.0, -K0_INITIAL, 0.0, 500, 0.0)
            .expect("valid RF single-ray trace parameters");

        // Ray should propagate inward
        assert!(result.trajectory.len() > 1, "Should have trajectory points");

        // First point at antenna
        let (r0, _z0) = result.trajectory[0];
        assert!((r0 - R_ANTENNA).abs() < 1e-10, "Should start at antenna");
    }

    #[test]
    fn test_trace_rays_array() {
        let rf = RFHeatingSystem::new();
        let results = rf
            .trace_rays()
            .expect("valid RF multi-ray trace parameters");
        assert_eq!(results.len(), N_RAYS);
        for result in &results {
            assert!(!result.trajectory.is_empty());
        }
    }

    #[test]
    fn test_alfven_speed_positive() {
        let rf = RFHeatingSystem::new();
        let v_a = rf.alfven_speed(6.2, 0.0);
        assert!(
            v_a > 0.0 && v_a.is_finite(),
            "v_A should be positive: {v_a}"
        );
    }

    #[test]
    fn test_rf_heating_rejects_invalid_runtime_inputs() {
        let mut rf = RFHeatingSystem::new();
        rf.b0 = f64::NAN;
        assert!(rf.resonance_radius().is_err());
        assert!(rf.trace_rays().is_err());

        let rf = RFHeatingSystem::new();
        assert!(rf
            .trace_single_ray(0.0, 0.0, -K0_INITIAL, 0.0, 500, 0.0)
            .is_err());
        assert!(rf
            .trace_single_ray(R_ANTENNA, 0.0, -K0_INITIAL, 0.0, 0, 0.0)
            .is_err());
    }
}
